strongbox-web-core/src/main/java/org/carlspring/strongbox/controller/ArtifactController.java

```diff
@@ -27,32 +17,50 @@
 import java.util.Locale;
 import java.util.regex.Matcher;
 
-import io.swagger.annotations.ApiOperation;
-import io.swagger.annotations.ApiParam;
-import io.swagger.annotations.ApiResponse;
-import io.swagger.annotations.ApiResponses;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.xml.bind.JAXBException;
+
 import org.apache.commons.io.FileUtils;
 import org.apache.commons.io.comparator.DirectoryFileComparator;
 import org.apache.log4j.LogManager;
 import org.apache.log4j.Logger;
 import org.apache.maven.artifact.repository.metadata.Metadata;
+import org.carlspring.maven.commons.util.ArtifactUtils;
+import org.carlspring.strongbox.client.ArtifactTransportException;
+import org.carlspring.strongbox.io.ArtifactInputStream;
+import org.carlspring.strongbox.providers.ProviderImplementationException;
+import org.carlspring.strongbox.storage.ArtifactResolutionException;
+import org.carlspring.strongbox.storage.ArtifactStorageException;
+import org.carlspring.strongbox.storage.Storage;
+import org.carlspring.strongbox.storage.metadata.MavenMetadataManager;
+import org.carlspring.strongbox.storage.metadata.MetadataType;
+import org.carlspring.strongbox.storage.repository.Repository;
+import org.carlspring.strongbox.util.MessageDigestUtils;
 import org.codehaus.plexus.util.xml.pull.XmlPullParserException;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.MediaType;
 import org.springframework.http.ResponseEntity;
 import org.springframework.security.access.prepost.PreAuthorize;
-import org.springframework.web.bind.annotation.*;
-import static org.carlspring.strongbox.utils.ByteRangeRequestHandler.handlePartialDownload;
-import static org.carlspring.strongbox.utils.ByteRangeRequestHandler.isRangedRequest;
-import static org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR;
-import static org.springframework.http.HttpStatus.NOT_FOUND;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
 
 /**
  * test
  */
 @RestController
-@RequestMapping("/storages")
+@RequestMapping(path = ArtifactController.ROOT_CONTEXT, headers = "user-agent=Maven/*")
 public class ArtifactController
```

Comment:
Earlier we decided that each `Protocol` (Maven, Nuget, etc.) can have multiple layouts (Hierarchical, Plane etc.) which is controlled by `LayoutProvider`, and `path` means the layout path,  that's from one side. From the other side, when we want to store artifact (package) in `Controller`, we don't know which `Layout` we have under `Storage` and `Repository` and we can't build its `Path`. So we need some kind of abstraction which will point the artifact in `Repository` and `ArtifactCoordinates` can be such type of abstraction. But you are right about `toPath()` method and in this case we must move toPath() method from `ArtifactCoordinates` to `LayoutProvider.toPath(ArtifactCoordinates)`.